### PR TITLE
tokenizer migrated to g2p from ReadAlongs/Studio

### PR DESCRIPTION
**Pull request by joanise**
`g2p convert` now accepts a `--tok` switch to tokenize before converting:
 - `--tok` tokenizes according to the path from `in_lang` to `out_lang`, stopping after two hops or the first -ipa node
 - `--tok-lang LANG` tokenizes instead according to the path from LANG to LANG-ipa, up to two hops (can be different from `in_lang`)
 - `--no-tok` turns of tokenization (default)

`make_g2p()` now accepts an optional `tok_lang` parameter
 - `tok_lang="path"` will tokenize according to the path from `in_lang` to `out_lang`, up to two hops or the first -ipa node
 - `tok_path="LANG"` will tokenize according to the path from LANG to LANG-ipa, up to two hops
 - `tok_path=None` (default) no tokenization

Additional changes:
 - fixed the tokenizer to handle a lot more situations correctly than it did in Studio.
 - `tli` ejective marker now handled so it tokenizes correctly
 - hack for recognizing the period as a letter in `tli` when not word final - this code needs to be made more general, but for now it's enough so that all the unit tests pass
 - The unit test `test_cli.py` runs all examples in `public/data` with and without `--tok`: it is considered an error in the mapping itself if a word is mapped differently when tokenization is enabled. That would normally mean it needs to explicitly map some non-letter character, like the ejective marker for `tli`.

### CODE REVIEW

File: g2p/mappings/tokenizer.py [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+            if in_lang == "tli":
+                self.tokenizers[tokenizer_key].dot_is_letter = True
+
+        return self.tokenizers.get(tokenizer_key)
+
+
+_TOKENIZER_LIBRARY = TokenizerLibrary()
+
+
+def get_tokenizer(in_lang=None, out_lang=None, tok_path=None):
```
Comment (roedoejet) [inline g2p/mappings/tokenizer.py:186]: Nice. This is clear, and a very reasonable strategy I think.

**Review by roedoejet — APPROVED** [review]
Great work @joanise ! This will be a really valuable feature in `g2p`. You've thought of a lot of edge cases here and test coverage is thorough. I made some small changes and minor suggestions, but I'm happy to merge as-is if you have a peek at the comments.

File: g2p/transducer/__init__.py [hunk 2]
```diff
@@ -491,7 +592,7 @@ class CompositeTransducer():
 
     def __init__(self, transducers: List[Transducer]):
```
[the remark this answers is not in the document]
Comment (joanise) [reply g2p/transducer/__init__.py:593]: TokenizingTransducer can consume CompositeTransducers, but not vice versa.

File: g2p/transducer/__init__.py [hunk 3]
```diff
@@ -119,21 +123,56 @@ def tiers(self):
...
+            "output": self._output_string,
+            "input_nodes": self._input_nodes,
+            "output_nodes": self._output_nodes,
+        }
 
-class Transducer():
+    def clear_debugger(self):
+        self._debugger = []
+
+    def append(self, tg):
```
[the remark this answers is not in the document]
Comment (joanise) [reply g2p/transducer/__init__.py:154]: Oh boy, I'm not familiar with those overloads. I'd have to review the code and see if that would make the logic complete, but yeah, it sounds like a good idea. append() is pretty narrow, though, you can only append a "compatible" tg's: a basic one, or a composite one with the same tiers.

File: g2p/mappings/tokenizer.py [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+                    mapping = Mapping(in_lang=in_lang, out_lang=out_lang)
+                    self.tokenizers[tokenizer_key] = Tokenizer(mapping)
+                except MappingMissing:
+                    self.tokenizers[tokenizer_key] = self.tokenizers[None]
+                    LOGGER.warning(
+                        f"Cannot find mapping from '{in_lang}' to '{out_lang}'. Using default tokenizer instead!!!"
+                    )
+
+            # Hack for Tlingit using dot as a letter when non word-final
+            if in_lang == "tli":
```
[the remark this answers is not in the document]
Comment (joanise) [reply g2p/mappings/tokenizer.py:177]: Yes, that's what we should do. But so far this is the only language that needs it. I would like to leave it as is for now, and factor it out appropriately once we figure out how to do it. The bigger picture is that I'd like to be able to say "this punctuation is a letter in this or that context" for any punctuation. e.g., maybe ' is only a letter after a stop, or in the middle of the word. Here . is only a letter when not word final. I want to design a better solution, but I'd like that made a separate issue we can solve outside this PR. I just created issue #86 to that effect.

File: g2p/mappings/tokenizer.py (hunk 1 above)
Comment (joanise) [reply g2p/mappings/tokenizer.py:186]: Thanks, I'm glad you like it. It took me a while to get something I was happy with!

File: g2p/transducer/__init__.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (joanise) [reply g2p/transducer/__init__.py:154]: Hum, I just did some reading on this, and __iadd__ is wonky on complex objects. Apparently, 

    x+=y

is effectively translated to 

    x = x.__iadd__(y)

Ref: https://stackoverflow.com/a/31546707/3216427

this means:
 - __iadd__() should end with `return self`
 - the object reference gets copied around - I guess that's not a deep copy, so it's probably fine

[the remark this answers is not in the document]
Comment (joanise) [reply g2p/transducer/__init__.py:154]: Well, how weird, my code is every so slightly faster when I use

    tg += word_tg
    tg += non_word_tg

than

    tg.append(word_tg)
    tg.append(non_word_tg)

I'll commit the change!

**Comment by joanise** [comment]
Thanks for the review. Changes accepted, and suggestion for tg1 += tg2 implemented.
